### DISCUSSION TITLE
Image in Table Preventing Table Extraction

### DISCUSSION THREAD

**Discussion by isaac-peter** [pymupdf/PyMuPDF#3585] [Looking for help] [answered]
Hello,

On page 7 of the attached pdf, you will notice that two small images with links appear in the table with first column "Last Raced". When I run:
doc = pymupdf.open('eqbPDFChartPlus - 2024-06-10T140626.277.pdf') # Open document
doc_page = doc[6]
tabs = doc_page.find_tables(horizontal_strategy="text", vertical_strategy="text", clip = [9.920000076293945, 130.47003173828125, 180.35995483398438, 152.72203063964844], )

I get an error message indicating that a table doesn't exist.

I don't have this issue with other pdfs in the same document (clip slightly different) but which do not contain a small image in the table.

Any thoughts on how to extract the table with the first column titled "Last Race"?

I tried creating a new pdf with the link removed, but PyMuPDF still did not grab the table.

Thanks!

[eqbPDFChartPlus - 2024-06-10T140626.277.pdf](https://github.com/user-attachments/files/15849179/eqbPDFChartPlus.-.2024-06-10T140626.277.pdf)

**Comment by JorjMcKie** [comment]
Sorry, this unfortunately is one of the cases, where PyMuPDF table recognition fails: no grid lines whatsoever, unclear word positioning information, etc.
The presence of these little image has no influence.

**Reply by isaac-peter** [reply]
Hello,

I appreciate the prompt reply.

Two follow up questions.

1) Do you have any other methods you could suggest to extract that data will keeping column/row alignment?

2) Would it be worth trying to OCR the PDF, and then trying to find the table? If so, do you have a recommended OCR engine?

Thanks again!

**Reply by JorjMcKie** [reply]
Ad 2.: **_Definitely not._** OCR always means that you **_lose_** information. All text information is already available and will Only become worse under OCR. In addition any OCR engine **_cannot_** know anything about the original font, font size etc. It also usually does not extract vector graphics. So you won't ever have cell gridlines.

Ad 1.:
The more information you can add to the table recognizer, the better. In your special case, you could provide a clip bbox inside which your table is located.
The table finder accepts a list of "virtual lines" which it uses as if they were drawn vector graphic lines. Each virtual line is a pair of points (or point-likes).

Here is an approach to build a list of virtual lines for your desired table:
1. Locate string "Last Raced" via `rect1 = page.search_for("Last Raced")[0]`. The top-left of this rectangle also is the top-left of the desired clip.
2. `rect2 = page.search_for("Fractional Times:")[0]`. Is top coordinate rect2.y0 is the bottom coord of clip.
3. So define `clip = pymupdf.Rect(rect1.tl, page.rect.width, rect2.y0)`. Inside this rectangle is your desired table and nothing else. Start building up your list `add_lines` by adding the 4 border lines of the clip rectangle:
    ```python
    add_lines = [(clip.tl, clip.tr), (clip.tr, clip.br),
                 (clip.br, clip.bl), (clip.bl, clip.tl)]
    ```
5. Now find the x-coordinates of the columns. Do this by searching for the strings "Pgm", "Horse Name (Jockey)", "Wgt", etc. For each of the returned search hits, take the x0: these are the column borders (11 in total). For each x in this list do this to add column border lines to the virtual lines:
    ```python
    add_lines.append((x, clip.y0), (x, clip.y1))
    ```
7. At this point, only the y-coordinates of the table rows are missing. This is a bit of a challenge, because we have so many superscripts in the table, that we need to use logic that still finds its way through that jungle. Luckily, we have package pymupdf4llm. This contains a helper function which does that for us and delivers cleaned-up rectangles for each line
    ```python
    import pymupdf4llm
    from pymupdf4llm.helpers.get_text_lines import get_raw_lines
    textpage = page.get_textpage()
    raw_lines = get_raw_lines(textpage, clip=clip)
    for bbox, spans in raw_lines:
        add_lines.append((bbox.bl, bbox.br))
    ```

A long story. But if you now do `page.find_tables(clip=clip, add_lines=add_lines)` you should be able to always get your table.

**Reply by JorjMcKie** [reply]
Here is the script that does it:
```python
import pymupdf
import pymupdf4llm
from pymupdf4llm.helpers.get_text_lines import get_raw_lines

doc = pymupdf.open("test.pdf")
page = doc[6]

rect1 = page.search_for("Last Raced")[0]
rect2 = page.search_for("Fractional Times:")[0]
clip = pymupdf.Rect(rect1.tl, page.rect.width, rect2.y0)
add_lines = [
    (clip.tl, clip.tr),
    (clip.tr, clip.br),
    (clip.br, clip.bl),
    (clip.bl, clip.tl),
]

# names in the column header row
# these names vary somewhat by page, so we may want to
# add all variants
needles = (
    "Pgm",
    "Horse Name (Jockey)",
    "Wgt",
    "M/E",
    "PP",
    "Start",
    "1/4",
    "1/2",
    "Str",
    "Fin",
    "Odds",
    "Comments",
)

# from here on, we use a textpage to speed up searches and
# because we need one in get_raw_lines
textpage = page.get_textpage(clip=clip)

# determine column borders
for needle in needles:
    try:  # depending the page not all header names may exist
        r = page.search_for(needle, textpage=textpage)[0]
        add_lines.append(((r.x0, clip.y0), (r.x0, clip.y1)))
    except IndexError:
        pass

# determine row bottom coordinates using "raw" lines
# this a list of tuples (row-bbox, spans-in-row)
raw_lines = get_raw_lines(textpage, clip=clip)
for bbox, spans in raw_lines[:-1]:
    add_lines.append(((clip.x0, bbox.y1), (clip.x1, bbox.y1)))

# should now deliver all we want!
tabs = page.find_tables(clip=clip, add_lines=add_lines)
tab = tabs[0]
for r in tab.extract():
    print(r)
```

**Reply by isaac-peter** [reply]
Wow thank you so much for the detailed response! One quick follow-up question. Not all pages have 'Fractional Times:' at the bottom of the table; sometimes there is other text in bold. Do you have any other suggestions to mark the lower left side of the table? What about searching the document for the next Helvetica-Bold text that is not one of the needles? Then extracting the y0 value from that text box?

Thanks again!

**Reply by JorjMcKie** [reply, marked as answer]
> Wow thank you so much for the detailed response! One quick follow-up question. Not all pages have 'Fractional Times:' at the bottom of the table; sometimes there is other text in bold. Do you have any other suggestions to mark the lower left side of the table? What about searching the document for the next Helvetica-Bold text that is not one of the needles? Then extracting the y0 value from that text box?
> 
> Thanks again!

Certainly possible, but more complex. It would be easier if you knew a list of text alternatives and could check for them.
But you **_can_** extract all bold text spans and fight your way through this jungle.

```python
# bold text spans:
spans = [
    s
    for b in page.get_text("dict", flags=pymupdf.TEXTFLAGS_TEXT)["blocks"]
    for l in b["lines"]
    for s in l["spans"]
    if s["flags"] & 16  # font is flagged as "bold"
]

# sort them by y1, x0:
spans.sort(key=lambda s: (s["bbox"][3],s["bbox"][0]))

# find span containing "Last Raced":
top = [s for s in spans if "Last Raced" in s["text"]]

# get its bbox:
rect1 = pymupdf.Rect(top[0]["bbox"])

# now look for first span below rect1 (remember: they all are bold):
bot = [s  for s in spans if s["bbox"][3] > top_bbox[3]]
rect2 = pymupdf.Rect(bot[0]["bbox"])
```

**Reply by isaac-peter** [reply]
Great! Thank you again! This was really helpful!
